fasta/fasta.go

```diff
@@ -1,129 +1,160 @@
+	}
+
+	for curr, doneReading = NextFasta(file); !doneReading; curr, doneReading = NextFasta(file) {
+		if usedSeqNames[curr.Name] {
+			log.Fatalf("ERROR: %s is used as the name for multiple records. Names must be unique.", curr.Name)
 		} else {
-			currSeq = dna.StringToBases(line)
-			answer[seqIdx].Seq = append(answer[seqIdx].Seq, currSeq...)
+			usedSeqNames[curr.Name] = true
+			answer = append(answer, curr)
 		}
 	}
+
+	exception.WarningOnErr(file.Close())
 	return answer
 }
 
-func WriteToSplitChr(filename string, records []*Fasta) {
-	for _, rec := range records {
-		Write(filename+rec.Name+".fa", []*Fasta{rec})
+// NextFasta reads a single fasta record from an input EasyReader. Returns true when the file is fully read.
+func NextFasta(file *fileio.EasyReader) (*Fasta, bool) {
+	var line string
+	var name string
+	var answer *Fasta
+	var doneReading bool
+	var peek []byte
+	var err error
+
+	for line, doneReading = fileio.EasyNextRealLine(file); !doneReading; line, doneReading = fileio.EasyNextRealLine(file) {
+		if strings.HasPrefix(line, ">") {
+			name = line[1:]
+			answer = &Fasta{Name: name, Seq: make([]dna.Base, 0)}
+		} else {
+			if answer == nil {
+				log.Fatalf("ERROR: %s is missing a sequence name (e.g. >chr1)", file.File.Name())
+			}
+			answer.Seq = append(answer.Seq, dna.StringToBases(line)...)
+		}
+
+		peek, err = fileio.EasyPeekReal(file, 1)
+		if errors.Is(err, io.EOF) {
+			return answer, doneReading // we could return true here, but we do not in order to match the behavior of other Next functions.
+		} else {
+			exception.PanicOnErr(err)
+		}
+
+		if peek[0] == '>' {
+			return answer, doneReading
+		}
 	}
+	return answer, doneReading
 }
 
-func WriteToFileHandle(file io.Writer, records []*Fasta, lineLength int) {
-	var err error
-	for _, rec := range records {
-		_, err = fmt.Fprintf(file, ">%s\n", rec.Name)
-		common.ExitIfError(err)
-		for i := 0; i < len(rec.Seq); i += lineLength {
-			if i+lineLength > len(rec.Seq) {
-				_, err = fmt.Fprintf(file, "%s\n", dna.BasesToString(rec.Seq[i:]))
-				common.ExitIfError(err)
-			} else {
-				_, err = fmt.Fprintf(file, "%s\n", dna.BasesToString(rec.Seq[i:i+lineLength]))
-				common.ExitIfError(err)
-			}
+// ToMap converts the output of the Read function to a map of sequences keyed to the sequences name.
+func ToMap(ref []*Fasta) map[string][]dna.Base {
+	m := make(map[string][]dna.Base)
+	for i := range ref {
+		_, ok := m[ref[i].Name]
+		if !ok {
+			m[ref[i].Name] = ref[i].Seq
+		} else {
+			log.Panicf("%s used for multiple fasta records. record names must be unique.", ref[i].Name)
 		}
 	}
+	return m
 }
 
+// Write a fasta to input filename. Output fastas have line length of 50.
 func Write(filename string, records []*Fasta) {
 	lineLength := 50
 	file := fileio.EasyCreate(filename)
-	defer file.Close()
-
 	WriteToFileHandle(file, records, lineLength)
+	exception.WarningOnErr(file.Close())
```

Comment:
When I wrote this I thought the only possible error return is 'file already closed' which wasn't worth killing a program for. It's true that is the only error for os.File.Close() but I realize now that since EasyWriter wraps a gzip and bufio buffer, more serious errors are possible. To address this I have made some changes to the EasyWriter close function (I also found an important bug where I write it as gz OR bufio OR file where in reality it is AND) and changed the error check in the fasta package to panic on error instead of warning.